src/hazelcore/semantics/Action.re

```diff
@@ -3611,6 +3818,28 @@ module Exp = {
 
     /* No construction handled at block level */
 
+    /* SwapUp and SwapDown is handled at block level */
+    | (SwapUp, _) when ZExp.line_can_be_swapped(zline) =>
+      switch (ListUtil.split_last(prefix)) {
+      | None => Failed
+      | Some((rest, last)) =>
+        let new_zblock = (rest, zline, [last, ...suffix]);
+        Succeeded(
+          AnaDone(Statics.Exp.ana_fix_holes_z(ctx, u_gen, new_zblock, ty)),
+        );
+      }
+    | (SwapDown, _) when ZExp.line_can_be_swapped(zline) =>
+      switch (suffix, zline) {
+      | ([], _) => Failed
+      /* avoid swap down fot the Let line if it is second to last */
+      | ([_], LetLineZP(_) | LetLineZA(_) | CursorL(_, LetLine(_))) =>
+        Failed
+      | ([hd, ...tl], _) =>
+        let new_zblock = (prefix @ [hd], zline, tl);
```

Comment:
need to do a normalization pass to prevent ExpLines that consist of single holes from being included in the middle of a block (they should be turned into empty lines). There is already code for dealing with this situation somewhere (@dmoon1221 might have an answer for where)

Reply:
have done the normalization for all new_zblock